### PR TITLE
More e2e tests

### PR DESCRIPTION
**Pull request by dkoshkin**
**Is this a bug fix or adding new feature?**
/Feature
Fixes: #118 

**What is this PR about? / Why do we need it?**
Introduces many of the tests outlined in https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/118, the ones that are left are the `Failure Scenarios`

In this PR a few of the tests require EBS volumes to already exist, I've created a new `cloud` package in the e2e test folder that will Create/Delete volumes, unfortunately much of the code is duplicated from the `pkg/` directory but it didn't feel right trying to reuse that code in the test. 

**What testing is done?** 
```
Running Suite: AWS EBS CSI Driver End-to-End Tests
==================================================
Random Seed: 1548253558 - Will randomize all specs
Will run 30 of 30 specs

Running in parallel across 32 nodes

[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "io1" and fsType "ext4"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Pre-Provisioned
  [env] should write and read to a pre-provisioned volume
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/pre_provsioning.go:125
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should delete PV with reclaimPolicy "Delete"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:269
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "st1" and encryption
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:82
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "gp2" and encryption
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:82
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "io1" and encryption
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:82
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "gp2" and fsType "ext2"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [multi-az] Dynamic Provisioning
  [env] should allow for topology aware volume with specified zone in allowedTopologies
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:358
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "sc1" and fsType "ext2"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "sc1" and fsType "ext3"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Pre-Provisioned
  [env] should use a pre-provisioned volume and mount it as readOnly in a pod
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/pre_provsioning.go:149
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create multiple PV objects, bind to PVCs and attach all to different pods
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:166
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create multiple PV objects, bind to PVCs and attach all to different pods on the same node
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:204
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  [env] should retain PV with reclaimPolicy "Retain"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:286
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "io1" and fsType "ext2"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [multi-az] Dynamic Provisioning
  should allow for topology aware volume scheduling
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:331
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "gp2" and fsType "ext3"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create multiple PV objects, bind to PVCs and attach all to a single pod
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:133
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "st1" and fsType "ext4"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Pre-Provisioned
  [env] should use a pre-provisioned volume and delete PV with reclaimPolicy "Delete"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/pre_provsioning.go:191
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "st1" and fsType "ext3"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "st1" and fsType "ext2"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with provided mountOptions
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:108
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "sc1" and fsType "ext4"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "gp2" and fsType "ext4"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand and mount it as readOnly in a pod
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:244
•
------------------------------
[ebs-csi-e2e] [single-az] Pre-Provisioned
  [env] should use a pre-provisioned volume and retain PV with reclaimPolicy "Retain"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/pre_provsioning.go:174
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "io1" and fsType "ext3"
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:54
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a volume on demand with volumeType "sc1" and encryption
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:82
•
------------------------------
[ebs-csi-e2e] [single-az] Dynamic Provisioning
  should create a deployment object, write and read to it, delete the pod and write and read to it again
  /Users/dkoshkin/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/dynamic_provisioning.go:315
•
Ran 30 of 30 Specs in 535.110 seconds
SUCCESS! -- 30 Passed | 0 Failed | 0 Pending | 0 Skipped
```

### CODE REVIEW

**Comment by leakingtapan** [comment]
@dkoshkin could you update the test done section, the result is outdated with latest changes

File: tests/e2e/testsuites/dynamically_provisioned_delete_pod_tester.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/driver"
+	. "github.com/onsi/ginkgo"
+	"k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// DynamicallyProvisionedDeletePodTest will provision required StorageClass and Deployment
+// Testing if the Pod can write and read to mounted volumes
+// Deleting a pod, and again testing if the Pod can write and read to mounted volumes
+type DynamicallyProvisionedDeletePodTest struct {
```
Comment (leakingtapan) [inline tests/e2e/testsuites/dynamically_provisioned_delete_pod_tester.go]: I didn't find where do we run this test case, and im not sure why do we need to test it through deployment. For the sake of limiting the scope of this PR, can we hold off on this test case?

Comment (dkoshkin) [reply tests/e2e/testsuites/dynamically_provisioned_delete_pod_tester.go]: The file was probably collapsed, here is the test: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/173/files#diff-4b7e52cf33a96cc1c8a4d9974c34027aR315

I would prefer to keep this test as its actually pretty important, this is the only one thats tests the scenario of what happens when a pod restarts and the volume gets remounted, it tests the detach/attach flow and validates that the volume is not being formatted after the second attach.

**Comment by dkoshkin** [comment]
@leakingtapan after your LGTM, I'll squash. `gofmt` is all good, I've been doing that before pushing.

File: pkg/driver/driver.go [hunk 2]
```diff
@@ -48,7 +48,7 @@ type Driver struct {
 func NewDriver(endpoint string) (*Driver, error) {
 	klog.Infof("Driver: %v Version: %v", DriverName, driverVersion)
 
-	cloud, err := cloud.NewCloud()
+	cloud, err := cloud.NewCloud(nil)
```
Comment (leakingtapan) [inline pkg/driver/driver.go]: Could you create the metadata service and pass it in here? Im trying to avoid passing in nil.

**Comment by leakingtapan** [comment]
/lgtm

Have one minor comment about nil passing.

File: pkg/driver/driver.go (hunk 2 above)
Comment (dkoshkin) [reply pkg/driver/driver.go]: The problem with that is that its going to leak AWS implementation details into this driver code. If we are ok with that I can make the change.
Another problem is that `EC2Metadata` returned by `svc := ec2metadata.New(sess)` is also required in `&ec2rolecreds.EC2RoleProvider{Client: svc},`, passing in both also doesn't seem right. 
Maybe we can have a new function, something like below that is used in the e2e tests, and change back the original function `func NewCloud()  (Cloud, error)`:
```
func NewCloudWithMetadata(metadata MetadataService) (Cloud, error)
```

Comment (leakingtapan) [reply pkg/driver/driver.go]: `NewCloudWithMetadata` sounds a better choice for me. 

I feel the way we are handling metadata is a bit messy. Will need to refactor later to improve this.

Comment (dkoshkin) [reply pkg/driver/driver.go]: Sounds good, I'll make the change and test the code locally and push it up soon. But you are right, the metadata code needs some refactoring the interaction with `MetadataService` and `EC2Metadata` is confusing.

**Comment by dkoshkin** [comment]
Rebased onto the master and fixed the failing unit tests.

**Pull request merged**